src/radical/pilot/agent/launch_method/ibrun.py

```diff
@@ -41,38 +39,52 @@ def construct_command(self, cu, launch_script_hop):
 
         slots        = cu['slots']
         cud          = cu['description']
+
         task_exec    = cud['executable']
-        task_cores   = cud['cpu_processes']  # FIXME: handle cpu_threads
         task_args    = cud.get('arguments') or []
         task_argstr  = self._create_arg_string(task_args)
+        task_env     = cud.get('environment') or dict()
+
+        n_tasks      = cud['cpu_processes']  # FIXME: handle cpu_threads
 
-        cpn     = self._cfg.cores_per_node
-        index   = 0
-        offsets = list()
+        # Usage of env variable TACC_TASKS_PER_NODE is purely for MPI tasks,
+        #  and threads are not considered (info provided by TACC support)
+        n_node_tasks = int(task_env.get('TACC_TASKS_PER_NODE') or
+                           self._cfg.get('cores_per_node', 1))
+
+        if not self._node_list:
+            raise RuntimeError('attribute "_node_list" (from RM) is not set')
+        if not slots.get('nodes'):
+            raise RuntimeError('unit.slots.nodes is not set')
```

Comment:
I'm fine with the check, but I wonder if this case can legally happen.   The `NOOP` scheduler should not set `slots` in the first place, but then this test would result in an `AttributeError`.  All other schedulers should set `nodes` if they are able to place the task - and if not, the task would not arrive here.  Am I missing a case?

You test for `core_map` in the slots in the loop below, and I have the same question there.  Can that legally happen?  Of not, maybe use an assert so that we can optimize it away.  

The `NOOP` scheduler may not assign slots - in that case, IBRUN should still be able to run it, but we would leave placement to the batch systemin that case, and not specify `offsets` at all.  I don't think we need to cover it in this PR, but I wanted to mention it - maybe add a `FIXME`?

Reply:
I like idea with `assert`, will update accordingly

In general all that checks are of general purpose - if we use something then need to check that it has valid values, yeah, I didn't keep in mind the origins (didn't consider schedulers), I'll re-check it tho.

And now I think that with nodes-related parameters I did cut the possibility it could be just empty lists (?) Should I update it as:
`if self._node_list is not None` and `if slots.get('nodes') is not None` ?

Reply:
Great, thanks.  In general I think we should use asserts more frequently to assure our assumptions, I have been slowly introducing them in some code paths...